pkg/logs/sender/sender.go

```diff
@@ -6,220 +6,125 @@
+	stopChan    chan struct{}
 }
 
-// Sender contains all the logic to manage a stream of messages to destinations
-type Sender interface {
-	Start()
-	Stop()
-	Flush(ctx context.Context)
-}
-
-// SingleSender sends logs to different destinations.
-type SingleSender struct {
-	inputChan    chan *message.Message
-	outputChan   chan *message.Message
-	hasError     chan bool
+// Sender sends logs to different destinations.
+type Sender struct {
+	inputChan    chan *message.Payload
+	outputChan   chan *message.Payload
 	destinations *client.Destinations
-	strategy     Strategy
 	done         chan struct{}
-	lastError    error
-	trackErrors  bool
+	bufferSize   int
 }
 
-// NewSingleSender returns a new sender.
-func NewSingleSender(inputChan chan *message.Message, outputChan chan *message.Message, destinations *client.Destinations, strategy Strategy) *SingleSender {
-	return &SingleSender{
+// NewSender returns a new sender.
+func NewSender(inputChan chan *message.Payload, outputChan chan *message.Payload, destinations *client.Destinations, bufferSize int) *Sender {
+	return &Sender{
 		inputChan:    inputChan,
 		outputChan:   outputChan,
-		hasError:     make(chan bool, 1),
 		destinations: destinations,
-		strategy:     strategy,
 		done:         make(chan struct{}),
-		trackErrors:  false,
+		bufferSize:   bufferSize,
 	}
 }
 
 // Start starts the sender.
-func (s *SingleSender) Start() {
+func (s *Sender) Start() {
 	go s.run()
 }
 
 // Stop stops the sender,
 // this call blocks until inputChan is flushed
-func (s *SingleSender) Stop() {
+func (s *Sender) Stop() {
 	close(s.inputChan)
 	<-s.done
 }
 
-// Flush sends synchronously the messages that this sender has to send.
-func (s *SingleSender) Flush(ctx context.Context) {
-	s.strategy.Flush(ctx)
-}
+func (s *Sender) run() {
+	reliableDestinations := buildDestinationStates(s.destinations.Reliable, s.outputChan, s.bufferSize)
 
-func (s *SingleSender) run() {
-	defer func() {
-		s.done <- struct{}{}
-	}()
-	s.strategy.Send(s.inputChan, s.outputChan, s.send)
-}
+	sink := additionalDestinationsSink(s.bufferSize)
+	additionalDestinations := buildDestinationStates(s.destinations.Additionals, sink, s.bufferSize)
+
+	for payload := range s.inputChan {
 
-// send sends a payload to multiple destinations,
-// it will forever retry for the main destination unless the error is not retryable
-// and only try once for additional destinations.
-func (s *SingleSender) send(payload []byte) error {
-	for {
-		err := s.destinations.Main.Send(payload)
-		if err != nil {
-			if s.trackErrors && s.lastError == nil {
-				s.hasError <- true
+		sent := false
+		for !sent {
+			for _, destState := range reliableDestinations {
+				if !destState.destination.GetIsRetrying() {
+					destState.input <- payload
+					sent = true
+				}
 			}
-			s.lastError = err
 
-			metrics.DestinationErrors.Add(1)
-			metrics.TlmDestinationErrors.Inc()
-			if _, ok := err.(*client.RetryableError); ok {
+			if !sent {
```

Comment:
`sent` will always be false if there are no reliable destinations.  Is that a valid configuration?  If not, something here should probably check that invariant.

Reply:
Nope. Main is always a reliable destination and must exist when the logs agent is configured.